[1.21.10] Allow registration of custom debug screen entries

ApexModder opened the pull request:
This PR allows the registration of custom `DebugScreenEntry`s and including them into vanillas debug profiles, using the newly introduced `RegisterDebugEntriesEvent` which is fired on the mod event bus.

With the introduction of custom debug entries to keep them uniquely identifiable and readable on screen the rendering for them has been updated
- Modded entries render as full `ResourceLocation`s rather than just the `path` element
- String rendering has been updated to use `drawScrollingString`, this is to ensure that names that go out of bounds are still fully readable

<details>
<summary>In game example</summary>

[In game example](https://github.com/user-attachments/assets/ba2e41d6-039d-4a6a-b893-a625ca3e3924)

</details>

ApexModder committed:
Allow registration of custom debug entries

ApexModder committed:
Display full RL for modded entries

Render entry names as scrolling strings

XFactHD reviewed (commented):
If we're adding an API for this then we need to also decide what to do with `CustomizeGuiOverlayEvent.DebugText` (I personally would be in favor of deleting it unless anyone can come up with a use case that isn't covered by a custom debug entry) and then delete the `DebugScreenOverlay` reject.

ApexModder commented:
Yeah im with xfact, from what i can see the vanilla debug entry system allows everything our `CustomizeGuiOverlayEvent.DebugText` event did, so dropping it makes sense while we are still in BC.

ApexModder committed:
Add basic javadoc to `RegisterDebugEntriesEvent.register`

Spinoscythe commented:
Is there no way to amend existing `DebugScreenEntry`s instead of creating a whole new one?

ApexModder commented:
as far as i know, not really due to them all being their own unique instances or in my test examples case a lambda.

You should however be able to append debug text to existing groups, just need to obtain the group id which is a simple RL, mojang stores them as static constants in each entries class they are private though

Mojang does this to group all their system info together, see `DebugEntrySystemSpecs`

Spinoscythe commented:
might be helpful to AT the group Id's here.

ApexModder committed:
AT all DebugScreenEntry GROUP keys to public

This should allow modders to append custom debug texts into these groups using `DebugScreenDisplayer.addToGroup`

ApexModder committed:
Tear out the old `CustomizeGuiOverlayEvent.DebugText` event

ApexModder committed:
Allow including debug entries into vanillas debug profiles

Spinoscythe commented:
> as far as i know, not really due to them all being their own unique instances or in my test examples case a lambda.
> 
> You should however be able to append debug text to existing groups, just need to obtain the group id which is a simple RL, mojang stores them as static constants in each entries class they are private though
> 
> Mojang does this to group all their system info together, see `DebugEntrySystemSpecs`

Could you put this in the javadoc of the event so that modders are aware of this.

ApexModder committed:
Add javadoc stating you can append to existing groups but not modify existing entries

XFactHD commented on src/client/java/net/neoforged/neoforge/client/event/RegisterDebugEntriesEvent.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @ApiStatus.Internal
+    public static void validateProfiles(RegisterDebugEntriesEvent event) {
+        // we delegate validation to its own method to allow people to call 'includeInProfile' before 'register'
+        var defaultError = validateProfile(event, DebugScreenProfile.DEFAULT);
+        var performanceError = validateProfile(event, DebugScreenProfile.PERFORMANCE);
+
+        // we delay throwing the error to allow both profiles to be validated in the same run
+        // we should still validate 'performance' if 'default' is invalid
+        if (defaultError != null && performanceError != null) {
+            var error = new IllegalStateException("Fatal errors occurred while validating DebugScreenProfiles");
+            error.addSuppressed(defaultError);
+            error.addSuppressed(performanceError);
+            throw error;
+        } else if (defaultError != null)
+            throw defaultError;
+        else if (performanceError != null)
+            throw performanceError;
+    }
+
+    @Nullable
+    private static IllegalStateException validateProfile(RegisterDebugEntriesEvent event, DebugScreenProfile profile) {
+        var profileMap = event.getProfileMap(profile);
+        var invalidIds = Sets.difference(profileMap.keySet(), event.entries.keySet());
+
+        if (!invalidIds.isEmpty()) {
+            var logger = LogManager.getLogger();
+            var msg = logger.getMessageFactory().newMessage("Found {} unregistered debug entries in profile: {}", invalidIds.size(), profile.getSerializedName());
+
+            logger.fatal(msg.getFormattedMessage());
+            logger.fatal("Unregistered debug entries: {}", invalidIds.stream().map(ResourceLocation::toString).collect(Collectors.joining(",", "[", "]")));
+
+            // throw in dev to ensure people are correctly registering their entries when including them in profiles
+            if (!FMLEnvironment.isProduction())
+                return new IllegalStateException(msg.getFormattedMessage());
+        }
+
+        return null;
+    }
```
Why are these static if they need an instance of the class they are called on?

ApexModder replied:
"work around" to not have this internal method show in code completions, when looking for methods on the event instance.

Can easily swap this out to a instance method of thats preferred, i just didnt want people typing `event.` and seeing `validateProfiles` as a auto complete option.

XFactHD replied:
I would prefer them to be instance methods. Other events with internal methods also have them as instance methods and it works just fine.

XFactHD commented on src/client/java/net/neoforged/neoforge/client/event/RegisterDebugEntriesEvent.java:
```diff
@@ -0,0 +1,129 @@
+            error.addSuppressed(defaultError);
+            error.addSuppressed(performanceError);
+            throw error;
+        } else if (defaultError != null)
+            throw defaultError;
+        else if (performanceError != null)
+            throw performanceError;
```
Please make the braces consistent, this mix of braces and no braces is awful to read

XFactHD commented on src/client/java/net/neoforged/neoforge/client/event/RegisterDebugEntriesEvent.java:
```diff
@@ -0,0 +1,129 @@
+        var invalidIds = Sets.difference(profileMap.keySet(), event.entries.keySet());
+
+        if (!invalidIds.isEmpty()) {
+            var logger = LogManager.getLogger();
+            var msg = logger.getMessageFactory().newMessage("Found {} unregistered debug entries in profile: {}", invalidIds.size(), profile.getSerializedName());
+
+            logger.fatal(msg.getFormattedMessage());
+            logger.fatal("Unregistered debug entries: {}", invalidIds.stream().map(ResourceLocation::toString).collect(Collectors.joining(",", "[", "]")));
```
Prefer using SLF4J via `LogUtils.getLogger()` and use a `StringBuilder` instead of the logger's message factory

Spinoscythe reviewed (commented):
Do we really need a switch for 2 cases?

ApexModder committed:
Refactor to use SLF4J

ApexModder committed:
Make braces consistent

ApexModder commented:
> Do we really need a switch for 2 cases?

Easier to read personally than a ternary but i could swap out for a ternary or 'if's if thats preferred

Spinoscythe commented:
I would say ternary is better

XFactHD commented:
A switch statement is perfectly adequate for this, a ternary unnecessarily reduces readability and "future-proofness" in case vanilla gets additional profiles in the future.

ApexModder committed:
Refactor `validateProfiles` to being an instance method rather than static

XFactHD reviewed (commented):
Looks good to me now, though I'd like to let this sit for a couple more days in case someone comes up with a use case that requires the removed event and cannot be handled with a custom entry.

ApexModder commented:
Been made aware of the following issues that should be resolved before merging
- [x] Search bar only searches the `path` component from entry ids
- [x] Entries are appearing mixed in with other modded and vanilla entries
  - They should maybe be sorted after vanilla (dont sort vanilla entries)
  - Also sorted to be grouped by namespace

ApexModder committed:
Better handle categories, searching and sorting

ApexModder committed:
Reformat code

XFactHD commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1151,4 +1157,33 @@ public static GpuDevice createGpuDevice(long window, int debugLevel, boolean syn
+        // - blank/empty string, accept everything
+        // - accept entries whose namespace/path match given text
+        DebugScreenEntries.allEntries().forEach((id, value) -> {
+            if (text.isBlank() || SharedSuggestionProvider.matchesSubStr(text, id.getNamespace()) || SharedSuggestionProvider.matchesSubStr(text, id.getPath())) {
```
Does this handle search text declaring both namespace and (partial) path (i.e `namespace:partial_or_full_path`) properly?
The implementation of `SharedSuggestionProvider.matchesSubStr()` and `SharedSuggestionProvider.filterResources(Iterable, String, Function, Consumer)` (latter uses the former) suggests that it wouldn't and would just return nothing.

ApexModder replied:
Yes, you can search `mymod` and only entries whose namespace or path contain `mymod` will show up, logic was copied from the command suggestion PR (#2522)

XFactHD replied:
I'm not asking about specifying only a namespace, I'm asking specifically about typing a string containing both a namespace **and** (part of) a path.

ApexModder replied:
Ah i see, no that does not work currently (something i didnt test tbh). typing `dummy0:` shows nothing but `dummy0` works as expected
Will look into fixing that

ApexModder committed:
Fix debug entry search failing when `:` is provided (for full id searrches)

ApexModder committed:
Sort debug entries by namespace (vanilla first)

ApexModder committed:
Use a Comparator for sorting instead of Multimap shenanigans

covers1624 commented on patches/net/minecraft/client/gui/components/debug/DebugScreenEntries.java.patch:
```diff
@@ -0,0 +1,36 @@
++        // Neo: Allow registering custom entries
++        net.neoforged.fml.ModLoader.postEventWithReturn(new net.neoforged.neoforge.client.event.RegisterDebugEntriesEvent(ENTRIES_BY_LOCATION, map, map1)).validateProfiles();
++        // Neo: Make the profile maps immutable again
++        PROFILES = Map.of(DebugScreenProfile.DEFAULT, Map.copyOf(map), DebugScreenProfile.PERFORMANCE, Map.copyOf(map1));
```
I recommend leaving the local variables immutable, passing them to the event wrapped in `new HashMap` then construct this by pulling the maps from the event + `Map.of`. This would reduce the patch size above.

ApexModder replied:
While that would make the patch smaller, i not a fan of the profile maps being publicly visible and mutable in the event.
This would allow people to listen on lowest priority and wipe out not only vanilla but other mods inclusions into them, by calling `event.getProfileMap(<profile>).clear()`

XFactHD commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1151,4 +1158,51 @@ public static GpuDevice createGpuDevice(long window, int debugLevel, boolean syn
         return glDevice;
     }
+
+    public static final Comparator<ResourceLocation> CMP_BY_NAMESPACE_VANILLA_FIRST = Comparator
```
Minor nitpick: I would move this to the top of the class (above the screen layer list) since this is usable for more than just sorting the debug entries, it's re-usable for any case where vanilla-first RL sorting is desired.

ApexModder replied:
If we are moving it up so other people can use it, maybe should also be moved into `CommonHooks` so that its not client only, since the comparator can safely be used on either side.

XFactHD replied:
Sure, sounds reasonable to me

ApexModder committed:
Not comparator instead of reversing it

ApexModder committed:
Move comparator into CommonHooks

ApexModder committed:
Move the profile map mutable patch down to make patch size smaller

ApexModder committed:
Simplify `DebugScreenEntries` patch to a 1-liner

XFactHD reviewed (approved):
After talking to someone who previously used `CustomizeGuiOverlayEvent.DebugText` (and brought up a good point) and checking references to the event in 1.21.1 via WAIFU, I'm confident that we're good to remove the event. As a replacement for certain use cases of the event, we may want to look into introducing something that is fired for each active debug entry to allow appending to it (adding a custom entry after a vanilla one is not an option due to sorting and the way vanilla selects which side an entry is displayed on), either via an "entry render event" or a means to register a "modifier" for a given entry. This can however be dealt with in a separate PR.
